### PR TITLE
 [GEOS-10142] OpenSearch for EO, switching Atom HTML description to Freemarker templates

### PR DESCRIPTION
**Pull request by aaime**
[GEOS-10142](https://osgeo-org.atlassian.net/browse/GEOS-10142)

See https://osgeo-org.atlassian.net/browse/GEOS-10142

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.**

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the `main` branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [ ] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [ ] PR for bug fixes and small new features are presented as a single commit
- [ ] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [ ] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [ ] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by Continuous Integration after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by Continuous Integration after opening this PR)
- [ ] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates (screenshots, text)
- [ ] Commits changing the REST API, or any configuration object, should check if the REST API docs (Swagger YAML files and classic documentation) need to be updated.

### CODE REVIEW

**Review by dromagnoli — APPROVED** [review]
Looks good to me

File: doc/en/api/1.0.0/opensearch-eo.yaml [hunk 1]
```diff
@@ -644,7 +559,7 @@ paths:
     get:
       description: |
         Retrieves a product search attributes, ogc links and granules.
-        The response contains cross links to ogcLinks/metadata/description/thumnail/granules sub-resources, when the same JSON is used o create a new collection the associated can be omitted, or if everything needs to be put in the JSON, the associated representations can be inlined instead (the thumbnail image can be BASE64 encded). A creation with ZIP is recommend.
+        The response contains cross links to ogcLinks/metadata/thumbnail/granules sub-resources, when the same JSON is used o create a new collection the associated can be omitted, or if everything needs to be put in the JSON, the associated representations can be inlined instead (the thumbnail image can be BASE64 encded). A creation with ZIP is recommend.
```
Comment (bradh) [inline doc/en/api/1.0.0/opensearch-eo.yaml:562]:
```suggestion
        The response contains cross links to ogcLinks/metadata/thumbnail/granules sub-resources, when the same JSON is used to create a new collection the associated can be omitted, or if everything needs to be put in the JSON, the associated representations can be inlined instead (the thumbnail image can be BASE64 encoded). A creation with ZIP is recommend.
```

File: doc/en/user/source/community/opensearch-eo/configuration.rst [hunk 2]
```diff
@@ -84,6 +84,28 @@ It is possible to add new product classes as well as changing the build-in ones,
...
 remember to "reset" the GeoServer configuration to make it re-scan the table structures.
 
+.. _oseo_html_templates:
+
+HTML templates
+--------------
+
+The Freemarker templates for collections and products can be found in ``$GEOSERVER_DATA_DIRECTORY/templates/os-eo``:
+
+* `product.ftl <https://github.com/geoserver/geoserver/blob/main/src/community/oseo/oseo-service/src/main/resources/org/geoserver/opensearch/eo/response/product.json>`_ receives one OpenSearch product and encodes the Atom HTML description for it.
```
Comment (bradh) [inline doc/en/user/source/community/opensearch-eo/configuration.rst:94]:
```suggestion
* `product.ftl <https://github.com/geoserver/geoserver/blob/main/src/community/oseo/oseo-service/src/main/resources/org/geoserver/opensearch/eo/response/product.json>`_ receives an OpenSearch product and encodes the Atom HTML description for it.
```

File: doc/en/user/source/community/opensearch-eo/configuration.rst [hunk 3]
```diff
@@ -84,6 +84,28 @@ It is possible to add new product classes as well as changing the build-in ones,
...
+* `collection.ftl <https://github.com/geoserver/geoserver/blob/main/src/community/oseo/oseo-service/src/main/resources/org/geoserver/opensearch/eo/response/collection.json>`_ receives on OpenSearch collection and encodes the Atom HTML description for it.
```
Comment (bradh) [inline doc/en/user/source/community/opensearch-eo/configuration.rst:95]:
```suggestion
* `collection.ftl <https://github.com/geoserver/geoserver/blob/main/src/community/oseo/oseo-service/src/main/resources/org/geoserver/opensearch/eo/response/collection.json>`_ receives an OpenSearch collection and encodes the Atom HTML description for it.
```

File: doc/en/user/source/community/opensearch-eo/upgrading.rst [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+Removal of ``htmlDescription``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Starting with version 2.20 the OpenSearch module dropped the HTML template columns in the
+database, and switched to freemarker templates instead. This relieves the database from a 
+significant burden, especially on the products table.
+
+The default templates are automatically used, and the old ``htmlDescription`` columns ignored
+(they should thus be removed).
```
Comment (bradh) [inline doc/en/user/source/community/opensearch-eo/upgrading.rst:14]:
```suggestion
(they should therefore be removed). 
```

File: doc/en/user/source/community/opensearch-eo/upgrading.rst [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+Starting with version 2.20 the OpenSearch module dropped the HTML template columns in the
+database, and switched to freemarker templates instead. This relieves the database from a 
+significant burden, especially on the products table.
+
+The default templates are automatically used, and the old ``htmlDescription`` columns ignored
+(they should thus be removed). 
+
+In order for the default `collection.ftl <https://github.com/geoserver/geoserver/blob/main/src/community/oseo/oseo-service/src/main/resources/org/geoserver/opensearch/eo/response/collection.json>`_
+to work two new fields, ``title`` and ``description``, should be added to the database
```
Comment (bradh) [inline doc/en/user/source/community/opensearch-eo/upgrading.rst:17]:
```suggestion
to work, two new fields, ``title`` and ``description``, should be added to the database
```

File: doc/en/user/source/community/opensearch-eo/upgrading.rst [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+structure, if not already present.
+
+As a result of these changes, the REST resources previously used to manage the description templates
+have been removed, and eventual HTML description templates included in product or collection
```
Comment (bradh) [inline doc/en/user/source/community/opensearch-eo/upgrading.rst:21]:
```suggestion
have been removed, and residual HTML description templates included in product or collection
```

(or just leave it out)

**Pull request merged**